ENH: function to aggregate staypoints horizontally

NinaWie opened the pull request:
closes #257 

I added a function in preprocessing of staypoints where staypoints are merged if they are at the same location id and temporally less than a threshold apart.

Open questions:
- Issue #257 is about "aggregation / segmentation", but I don't understand so far how staypoints could be segmented?
- Probably, it would be good if this function again is not dependent on the locations column. Should I maybe set a parameter `max_dist` again to use the `geom` column for aggregating, if the `location_id` column is not available?
- Issue #257 sounds more general, i.e. not only about staypoints. How would a merge of triplegs look like?

hong2223 commented:
for the first point: splitting I think refers to temporally splitting, as we do in the [_split_overlaps](https://github.com/mie-lab/trackintel/blob/560c5f6e29cebc6ff84424c7f0f3d85eadac1e27/trackintel/analysis/tracking_quality.py#L206) function to split records that span several days. So far I can only think of that use case of `splitting` and I think you do not need to add extra functions to this PR

hong2223 commented:
for the second and third points (sorry for splitting the comments):

> Probably, it would be good if this function again is not dependent on the locations column. Should I maybe set a parameter max_dist again to use the geom column for aggregating, if the location_id column is not available?

- IMO, doing this is also a spatial clustering of staypoints as we do when generating locations, except that we do not assign an id to the result. So I think this might not be necessary and we should require location_id for the function.

> Issue Support for horizontal aggregation/segmentation #257 sounds more general, i.e. not only about staypoints. How would a merge of triplegs look like?

- To be honest, I also do not know how to merge triplegs. triplegs might be merged if they are connected spatiotemporally, but the travel mode also plays a role: we might be able to merge two triplegs with mode walk, but two triplegs with mode bus might indicate the person is travelling by two bus lines, which should not be merged. So I'd say we leave tripleg merging until we find a use case.

hong2223 commented:
Just a comment:

I am thinking of the cases in #257 and our processing pipeline (we generate staypoints from pfs and regard all the other pfs as triplegs). Setting the `max_time_gap` in `merge_staypoints(stps, max_time_gap="10min")` larger than 0 would potentially merge stps1-tpls-stps2 sequences into a single stps1, which is then not backward compatible, i.e., the correspondences of pfs-stp and pfs-tpls are now wrong. Do we need extra steps (e.g. tpls should be deleted and pfs of both tpls and stps2 assigned to stps1) to ensure compatibility or what is the use case of such a function?

henrymartin1 commented:
> Open questions:
> 
>     * Issue [Support for horizontal aggregation/segmentation #257](https://github.com/mie-lab/trackintel/issues/257) is about "aggregation / segmentation", but I don't understand so far how staypoints could be segmented?
As @hong2223 said. So I would only focus on aggregation for staypoints for now. 

>     * Probably, it would be good if this function again is not dependent on the locations column. Should I maybe set a parameter `max_dist` again to use the `geom` column for aggregating, if the `location_id` column is not available?
Sure, why not. If it is not too much work, it would be a plus to include this. 

>     * Issue [Support for horizontal aggregation/segmentation #257](https://github.com/mie-lab/trackintel/issues/257) sounds more general, i.e. not only about staypoints. How would a merge of triplegs look like?
Hm... for triplegs the problem is much more complicated as. Especially if the solution should be generic. Triplegs should by definition only contain a single continuous transport mode. So if we would have a way to determine that two consecutive triplegs were both of mode walk we could merge them. In practice however, this information is hard to get only based on the movement data itself. 
For splitting it is a bit easier so we did for example compare public transport triplegs with the public transport network and schedule and could identify where to split the tripleg. 


> 
> 
> Just a comment:
> 
> I am thinking of the cases in #257 and our processing pipeline (we generate staypoints from pfs and regard all the other pfs as triplegs). Setting the `max_time_gap` in `merge_staypoints(stps, max_time_gap="10min")` larger than 0 would potentially merge stps1-tpls-stps2 sequences into a single stps1, which is then not backward compatible, i.e., the correspondences of pfs-stp and pfs-tpls are now wrong. Do we need extra steps (e.g. tpls should be deleted and pfs of both tpls and stps2 assigned to stps1) to ensure compatibility or what is the use case of such a function?

Ah ok, interesting case. I don't think that every function has to be fully compatible end-to-end (e.g., from positionfixes to tours). Often it will depend on your definitions of staypoints. So if you say that you want to merge staypoints if there is only a 10 minute gap in between then that's what you get 🤷‍♂️

Hm... having potentially overlapping staypoints and triplegs are indeed be a problem for the trip generation. I am not really sure what to do about it. I would like to avoid going back to the positionfixes as it is probably a lot of work to perform this backwards assignment. 
The main use-cases that I have in mind comes from datasets such as GC1 where we work with only staypoints/triplegs (so not with positionfixes) and often have these chains staypoints with very small time gaps in between at the same location with the same activity level.

NinaWie commented:
Okay I'm not sure whether I can follow - should we discuss it in the next meeting? Is there anything I can do in the meantime on this PR? 
If you check the code, there are some other open questions coming up, e.g. which one to use as `activity`, `next_trip_id`, `trip_id` etc when aggregating - should these all be lists?

NinaWie commented:
I modified the function as discussed to take the triplegs as input as well and to only merge staypoints if there is not tripleg inbetween. I also added documentation saying that the pfs-stps relation is broken after staypoint aggregation, and I added some tests.
Should be ready to be reviewed again by @henrymartin1  or @hong2223

hong2223 reviewed (commented):
A clear and nice solution. I only have comments related to the understanding of the function.
Would also be nice if all the `stps` be renamed to `sp` :)

NinaWie commented on trackintel/preprocessing/staypoints.py:
```diff
@@ -219,3 +219,106 @@ def _generate_locations_per_user(user_staypoints, distance_metric, db, geo_col):
+    """
+
+    if isinstance(max_time_gap, str):
+        max_time_gap = pd.to_timedelta(max_time_gap)
```
Thanks for the hint, I raise an error now and also added an test for it

NinaWie commented on trackintel/preprocessing/staypoints.py:
```diff
@@ -219,3 +219,106 @@ def _generate_locations_per_user(user_staypoints, distance_metric, db, geo_col):
+    # get information whether the there is a tripleg after a staypoint
+    stps_tpls["next_type"] = stps_tpls["type"].shift(-1)
+    # get only staypoints, but with next type information
+    stps_merge = stps_tpls[stps_tpls["type"] == "staypoint"]
```
Added comment above

NinaWie commented on trackintel/preprocessing/staypoints.py:
```diff
@@ -219,3 +219,106 @@ def _generate_locations_per_user(user_staypoints, distance_metric, db, geo_col):
+        Maximum duration between staypoints to still be merged.
+        If str must be parsable by pd.to_timedelta.
+
+    agg: dict, optional
```
I changed it, but not sure if the explanation is okay now. Maybe you can double check this :)

NinaWie commented on trackintel/preprocessing/staypoints.py:
```diff
@@ -219,3 +219,106 @@ def _generate_locations_per_user(user_staypoints, distance_metric, db, geo_col):
+    agg_dict.update(agg)
+
+    # aggregate values
+    stps_merged = stps_merge.groupby(by="index_temp").agg(agg_dict).sort_values(by=["user_id", "started_at"])
```
It's dropped automatically because of the `sp = sp.set_index(index_name)` I think. So it should be fine like this

NinaWie commented on trackintel/preprocessing/staypoints.py:
```diff
@@ -219,3 +219,106 @@ def _generate_locations_per_user(user_staypoints, distance_metric, db, geo_col):
+
+    Examples
+    --------
+    >>> stps.as_staypoints.merge_staypoints(tpls, max_time_gap="1h", agg={"geom":"first})
```
Ah I saw too late that I can simply click "Commit suggestion". I changed that manually now

NinaWie commented:
Thanks for the suggestions, I changed these things! Hope it's better now, let me know what else is there to do. And sorry for all the e-mails - I forgot that it's better to answer in a review instead of commenting to each of your comments.

hong2223 commented:
> Thanks for the suggestions, I changed these things! Hope it's better now, let me know what else is there to do. And sorry for all the e-mails - I forgot that it's better to answer in a review instead of commenting to each of your comments.

Thanks for the fix 👍. I changed some typos and we are ready to merge once the tests have passed. Job done :).
